File: OES_analysis.py
# -*- coding: utf-8 -*-
"""
Created on Thu Jan 25 10:22:36 2024

@author: Lukas
"""

import matplotlib.pyplot as plt
import glob
import numpy as np
from scipy.optimize import least_squares
from scipy.signal import find_peaks
import pandas as pd


def ptest_lorentzian(n_peaks, length=200, mean_I=1, mean_fwhm=2, mean_off=1):
    # mean_fwhm?
    x = np.arange(length)
    off = np.random.normal(mean_off, mean_off / 5)
    x0 = np.random.randint(0, length + 1, n_peaks)
    I = np.random.normal(mean_I, mean_I / 10, n_peaks)
    fwhm = np.random.normal(mean_fwhm, mean_fwhm / 5, n_peaks)
    params = np.concatenate(([off], x0, I, fwhm))
    y_data = SpectralProperties.multi_peak_func(x, params) + np.random.normal(0, 0.1, length)
    testing = FindPeaks(y_data)
    assert testing.i_x0 - x0 < fwhm


class SpectralProperties:
    def lorentzian(self, x, x0, a, gam):
        """
        Parameters
        ----------
        x : np array.
        x0 : peak position.
        a : amplitude.
        gam : width.

        Returns
        -------
        TYPE
            Lorentzian.
        """
        return a * gam ** 2 / (gam ** 2 + (x - x0) ** 2)

    def multi_peak_func(self, x, params):
        off = params[0]
        params = params[1:]
        assert not len(params) % 3
        n_peaks = int(len(params) / 3)
        params_x0 = params[:n_peaks]
        params_I = params[n_peaks:2 * n_peaks]
        params_fwhm = params[2 * n_peaks:]
        return off + sum([self.lorentzian(x, x0_i, I_i, fwhm_i) for x0_i, I_i, fwhm_i
                          in zip(params_x0, params_I, params_fwhm)])


class FindPeaks(SpectralProperties):
    def __init__(self, y_data, full_fit=False, height_factor=10, width_factor=1, prominence_factor=0):
        """

        :param y_data: np array of spectral amplitude data
        :param height_factor: conditions for recognising a peak in initial guess phase
        """
        self.y0 = y_data
        self.x_data = np.arange(len(y_data))
        self.y_ground = min(y_data)
        self.y_data = y_data - self.y_ground
        self.y_amp = max(self.y_data)
        self.y_data = self.y_data / self.y_amp
        height_req = np.average(np.abs(np.diff(self.y_data))) * height_factor
        width_req = len(self.y_data) / 1000 * width_factor
        prominence_req = width_req * prominence_factor
        self.i_guess = self._initial_guess(height_req, width_req, prominence_req)
        block_len = int((len(self.i_guess) - 1) / 3)
        self.i_off = self.i_guess[0]
        self.i_x0 = self.i_guess[1:block_len + 1]
        self.i_I = self.i_guess[block_len + 1:2 * block_len + 1]
        self.i_fwhm = self.i_guess[2 * block_len + 1:]
        if full_fit:
            i_popt = self.fit_peaks(self.i_guess)
            # warum hat self.i_popt nen self?
            self.i_off = i_popt[0]
            self.i_x0 = i_popt[1:block_len + 1]
            self.i_I = i_popt[block_len + 1:2 * block_len + 1]
            self.i_fwhm = i_popt[2 * block_len + 1:]

    def fit_peaks(self, guess):
        result = least_squares(self.res_multi_peak_func, x0=guess)
        popt = result.x  # order: off, n*x0,n*I, n*fwhm
        return popt

    def res_multi_peak_func(self, params):
        """
        Function to minimize

        Parameters
        ----------
        params : TYPE
            DESCRIPTION.

        Returns
        -------
        diff : TYPE
            DESCRIPTION.

        """
        diff = [self.multi_peak_func(x, params) - y
                for x, y in zip(self.x_data, self.y_data)]
        return diff

    def _initial_guess(self, height_req, width_req, prominence_req):
        """
        Guesses initial peak positions etc. The number of peaks is determined
        here. Makes or breaks the fit

        Parameters
        ----------
        height_req : TYPE
            DESCRIPTION.
        width_req : TYPE
            DESCRIPTION.
        prominence_req : TYPE, optional
            DESCRIPTION. The default is 0.

        Returns
        -------
        guess : TYPE
            DESCRIPTION.

        """
        # initial properties of peaks
        pk, properties = find_peaks(self.y_data, height=height_req,
                                    width=width_req, prominence=prominence_req)
        # extract peak heights and fwhm
        _I = properties['peak_heights']
        fwhm = properties['widths']

        # pack initial guesses together
        guess = np.concatenate(([self.y_ground], pk, _I, fwhm))
        return guess

    # def return_results(self, popt):
    #    """
    #    get the results into an easily accessible order. Note that off is
    #    returned last.
    #    """
    #    n_peaks = np.round((len(popt) - 1) / 3)
    #    off = popt[0]
    #    popt = popt[1:]
    #    x0 = popt[:n_peaks]
    #    I = popt[n_peaks:2 * n_peaks]
    #    fwhm = popt[n_peaks:2 * n_peaks]
    #    return x0, I, fwhm, off

    def plot_results(self, popt):
        resolution = 2000
        x_plot = np.linspace(0, max(self.x_data), resolution)

        fig = plt.figure(figsize=(10, 5))
        ax = fig.add_subplot(2, 1, 1)
        bx = fig.add_subplot(2, 1, 2)
        if len(popt) > 1:
            x0 = popt[1:int(len(popt) / 3) + 1]
            ax.plot(x0, self.multi_peak_func(x0, popt), 'o', ms=5)
            test_data = self.multi_peak_func(x_plot, popt)
            fit_data = self.y_ground + self.y_amp * self.multi_peak_func(x_plot, popt)
            ax.plot(x_plot, test_data, 'r--', lw=0.5)
            bx.plot(x_plot, fit_data)
        ax.plot(self.x_data, self.y_data, 'ok', ms=1)
        bx.plot(self.x_data, self.y0, ls='', marker='o', markersize=1)

        plt.show()


class Spectrum:
    def __init__(self, _y_data, _spectral_range):
        self.data = _y_data
        self.cleaned_data, self.uniformity = self.clean_data()
        self.cycles = self.extract_cycles()

    def clean_data(self):
        """
        Removes times at which the spectrum doesn't show sufficient variation and removes ground signal strength
        :return:
        """
        absolute_differences = np.abs(np.diff(self.data, axis=0))
        # check whether the variation of the data is sufficiently large compared to the total max. Should remove flat
        # periods
        threshold = np.max(self.data) * 10
        uniformity_check = np.sum(absolute_differences, axis=0) > threshold
        _data = self.data[:, uniformity_check]
        ground = np.min(_data, axis=0)
        _data = _data - ground
        return _data, uniformity_check

    def extract_cycles(self):
        """
        Splits up the data into cycles separated by dark periods and returns the average value of the spectrum in these
        cycles.
        :return:
        """
        # Find the indices separating activity from darkness by using the previous definition of uniformity
        indices = np.nonzero(self.uniformity[1:] != self.uniformity[:-1])[0] + 1
        cycle_data = np.split(self.cleaned_data, indices, axis=1)
        cycle_data = cycle_data[0::2] if self.uniformity[0] else cycle_data[1::2]
        # As periods may be of different length, they must be saved as a list and iterated over
        cycle_average = np.zeros((np.shape(data)[0], len(cycle_data)))
        for _i, cycle in enumerate(cycle_data):
            cycle_average[:, _i] = np.average(cycle, axis=1)
        return cycle_average

    def return_results(self, *params):
        return self.cleaned_data, self.cycles

    def peak_analysis(self, height_factor, width_factor):
        peak_table = []
        for i in range(np.shape(self.cycles)[1]):
            y_data = self.cycles[:, i]
            analysis = FindPeaks(y_data, height_factor=height_factor, width_factor=width_factor)
            df = pd.DataFrame(data={f'x0_{i}': analysis.i_x0, f'I_{i}': analysis.i_I})
            peak_table.append(df)
        return peak_table


names = glob.glob("Data/*.csv")
fig = plt.figure(figsize=(15, 15))

for image, name in enumerate(names):
    data = np.loadtxt(name, skiprows=24, delimiter=',', )
    extent_data = np.loadtxt(name, skiprows=24, usecols=1, delimiter=',')
    extent = 0, 800, np.min(extent_data), np.max(extent_data)
    data = data[4:]
    ax = fig.subplots()
    # ax = fig.add_subplot(3, 4, image+1)
    ax.set_ylabel("nm")
    ax.set_xlabel("t")
    spec = Spectrum(data)
    cleaned_data, cycle_averages = spec.return_results()
    peaks = spec.peak_analysis()
    ax.imshow(cleaned_data, norm='linear', origin='lower', aspect='auto')
    break

fig.tight_layout()
plt.show()

File: OES_analysis.jl
using Plots
using LsqFit
using Random
using Peaks
using DelimitedFiles
using DataFrames
using Profile
using CSV
using Statistics
using Flux
using BenchmarkTools
using Flux: crossentropy, Momentum
using CUDA
using BSON

function testdata_lorentzian(n_peaks, length=200, mean_I=1, mean_fwhm=2, mean_off=1)
    """
    Generate a random lorentzian spectrum modified by random noise with the given Parameters
    mean_fwhm: how broad the peaks are supposed to be
    """
    x_max = 100
    x = Vector{Float64}(range(0, x_max, length))
    off = (randn().+mean_off).*mean_off/5
    x0 = rand(1:x_max, n_peaks)
    I = (randn(n_peaks).+ 1).*mean_I .+ mean_I .+ 1
    fwhm = (randn(n_peaks).*mean_fwhm/3 .+mean_fwhm)
    # fwhm[fwhm .< 0] .= 0 WTF
    params = vcat(off, x0, I, fwhm)
    ydata = multi_peak_func(lorentzian, x, params) + randn(length).*0.1 .+ 1
    return ydata, x
    # testing = FindPeaks(y_data)
    # assert testing.i_x0 - x0 < fwhm
end

function lorentzian(x, x0, a, gam)
        """
        Parameters
        ----------
        x : np array.
        x0 : peak position.
        a : amplitude.
        gam : width.

        Returns
        -------
        TYPE
            Lorentzian.
        """
        return ((a * gam ^ 2) ./ ((x .- x0) .^ 2 .+ gam^2))
end

function multi_peak_func(shape::Function, x, params)
    """
    Sums over multiple peak shapes to get spectrum
    """
    off = params[1]
    # need to unpack or find prettier way to unpack list in for loop
    block_len = floor(Int, (length(params)-1)/3)
    params_x0 = params[2:block_len+1]
    params_I = params[block_len+2:2*block_len+1]
    params_fwhm = params[2*block_len+2:end]
    return off .+ sum([shape(x, i_params_x0, i_params_I, i_params_fwhm) for 
    (i_params_x0, i_params_I, i_params_fwhm) = zip(params_x0, params_I, params_fwhm)])
    
end

struct Spectrum{T}
    ydata::Vector{T}
    xdata::Vector{T}
end

function fit_peaks(func::Function,  data::Spectrum, guess)
    _multi_peak_func(x, params) = multi_peak_func(func, x, params)
    fit = curve_fit(_multi_peak_func, data.xdata, data.ydata, guess)
    return fit.param
end

function initial_guess(ydata, heightfactor, widthfactor, xdata)
    """
    achieve initial guess of peak positions and parameters for fit, using local maxima of the data 
    """
    min_prom = heightfactor * sum(abs.(diff(ydata)))/length(ydata)
    min_width = widthfactor
    pk, vals = findmaxima(ydata)
    pk, proms = peakproms!(pk, ydata; minprom=min_prom)
    pk, widths, _, _ = peakwidths!(pk, ydata, proms; minwidth=min_width)
    heights = ydata[pk]
    pk = xdata[pk]
    guess = vcat(minimum(ydata), pk, heights, widths)
    return guess
end

function plot_results(func, data::Spectrum, popt, yground, yamp)
    npixel = 2000
    x_plot = range(data.xdata[1], data.xdata[end], npixel)
    p1 = scatter(data.xdata, data.ydata, ms=0.8)
    p2 = scatter(data.xdata, yground .+ yamp .* data.ydata, ms=0.8)
    x0 = popt[2:Int((length(popt)-1)/3)+1]
    if length(popt) > 1    
        x0_fit = zeros(0)
        plot_fit = zeros(0)
        for x_i in x0
            push!(x0_fit, multi_peak_func(func, x_i, popt))
        end
        for x_i in x_plot
            push!(plot_fit, yground + yamp * multi_peak_func(func, x_i, popt))
        end
        scatter!(p1, x0, x0_fit)
        plot!(p2, x_plot, plot_fit, label="fit")
        display(plot(p1, p2, layout=(2,1), legend=false))
        return
    end
    display(plot(p1, p2, layout=(2,1)))
    return
end

function cleandata(data::Matrix; threshold=100)
    """
    Data has uniteresting down time, which gets removed due to its lower intensity
    Ground is also removed for better contrast in plot
    """
    ground = minimum(data, dims=1)
    _data = data .- ground
    n_observations = size(data)[2]
    uniformitycheck = vec(Bool.(zeros(n_observations)))
    BSON.@load "spec_recognition.bson" model
    features = zeros((4, n_observations))
    for (i, spec) in enumerate(eachcol(_data))
        features[:, i] = calculate_features(spec)
    end
    
    uniformitycheck = (model(features)[1,:] .> 0.5)
    
    
    _data = _data[:,uniformitycheck]
    
    return _data, uniformitycheck
end

function calculate_features(spec)
    return [mean(spec), maximum(spec), var(spec), exp(mean(log.(abs.(spec))))]
end

function AIbullshit()
    # spectrum_params: mean, max, var, geom_mean
    function generate_training_data(half_length::Int)
        active = rand(1:20, half_length)
        spec_length = 400
        x = zeros((4, 2*half_length))
        y = zeros(2*half_length)
        for (i, spec) in enumerate(active)
            active_spec, bin = testdata_lorentzian(spec, spec_length)
            passive_spec = randn(spec_length).*0.1 .+ randn(1)*0.2 .+ 1
            x[:,2*i] = calculate_features(active_spec)
            x[:,2*i-1] = calculate_features(passive_spec)
            y[2*i] = true
            y[2*i-1] = false
        end
        return x, y
    end
    x_train, y_train = generate_training_data(1000)
    x_test, y_test = generate_training_data(200)
    input_size = size(x_train, 1)  # Number of features in each spectrum
    output_size = 2  # Two classes: active and silent

    model = Chain(
        Dense(input_size, 30, relu),
        BatchNorm(30),
        Dense(30, 2),
        softmax
        )|> gpu;

    loss(x, y) = crossentropy(model(x), y)
    optim = Flux.setup(Flux.Adam(0.01), model)

    target = Flux.onehotbatch(y_train, [true, false])                   # 2×200 OneHotMatrix
    loader = Flux.DataLoader((x_train, target) |> gpu, batchsize=50, shuffle=true);
    for epoch in 1:1000
        Flux.train!(model, loader, optim) do m, x, y
            y_hat = m(x)
            Flux.crossentropy(y_hat, y)
        end
    end
    model = model |> cpu

    out2 = model(x_test)  # first row is prob. of true, second row p(false)

    print(mean((out2[1,:] .> 0.5) .== y_test))  # accuracy 94% so far!
    BSON.@save "spec_recognition.bson" model
end

function eval_spectralflatness(spectrum)
    # Stolen from ChatGPT, maybe debug
    # Avoid log(0) issues by adding a small constant
    eps = 1e-10

    # Calculate the geometric mean and arithmetic mean of the spectrum
    geo_mean = exp(mean(log.(spectrum .+ eps)))
    arith_mean = mean(spectrum)

    # Calculate spectral flatness
    flatness = geo_mean / arith_mean

    return flatness
end

function extractcycles(data, uniformitycheck)
    indices = findall(!iszero, uniformitycheck[2:end] .!= uniformitycheck[1:end-1]) .+ 1 # extract periods of falses and trues
    indices = vcat(1, indices, length(uniformitycheck))
    cycles = []
    for i in range(1, length(indices)-1)
        if indices[i+1] - indices[i] > 2
            if uniformitycheck[1]==1 && i%2!=0 # check whether the first or second block is the active one
                push!(cycles, data[:,indices[i]:indices[i+1]])
            elseif  uniformitycheck[1]==0 && i%2==0
                push!(cycles, data[:,indices[i]:indices[i+1]])
            end
        end
    end
    return cycles
end

function extractcycleaverages(data, uniformitycheck)
    """
    extracts averages of the active cycles. 

    data: the uncleaned data!
    """
    cycles = extractcycles(data, uniformitycheck)
    cycle_averages = vec(zeros(size(data)[1],0))  # dont know yet how many cycles are actually going to be realized
    for c in cycles
        if size(c)[2]>2
            vcat(cycle_averages, mean(c, dims=2)) # compute the rowwise average within one cycle
        end
    end
    return cycle_averages
end

function average_cycles(data, cycle_averages, index::Tuple{Int64, Int64})
    cycle = data[:,index[1]:index[2]]
    if size(cycle)[2]>2
        return hcat(cycle_averages, sum(cycle, dims=2)./size(cycle)[2]) # compute the rowwise average within one cycle
    end
    return cycle_averages
end


function peak_analysis(func::Function, cycles, xdata, heightfactor, widthfactor)
    peaktable=[]
    _multi_peak_func(x, params) = multi_peak_func(func, x, params)
    jacobian = (outjacin,p) -> ForwardDiff.jacobian!(outjacin, _multi_peak_func!, out, p)
    guess = initial_guess(cycles[:,1], heightfactor, widthfactor, xdata) # inital peak positions that get passed thorugh every fit. conserves peak number and hopefully assignment
    for i in range(1, size(cycles)[2])
        analysis = curve_fit(_multi_peak_func, jacobian, xdata, cycles[:, i], guess)
        analysis = analysis.param
        block_len = Int((length(analysis)-1)/3)
        plot_results(lorentzian, spec, analysis, 0, 1)
        df = DataFrame(x0 = analysis[2:block_len+1], I = analysis[block_len+2:2*block_len+1])
        push!(peaktable, df)
        break
    end
    return peaktable
end

# replace this some time
function peak_analysis_sparse(cycles, xdata, heightfactor, widthfactor)
    peaktable=[]
    spec1 = Spectrum(cycles[:, 1], xdata)
    # jacobian = (outjacin,p) -> ForwardDiff.jacobian!(outjacin, f!, out, p)
    guess = initial_guess(cycles[:,1], heightfactor, widthfactor, xdata) # inital peak positions that get passed thorugh every fit. conserves peak number and hopefully assignment
    for i in (1, size(cycles)[2])
        spec = Spectrum(cycles[:, i], xdata)
        analysis = fit_peaks(lorentzian, spec, guess)
        block_len = Int((length(analysis)-1)/3)
        order = sortperm(analysis[2:block_len+1])
        df = DataFrame(x0 = (analysis[2:block_len+1])[order], I = (analysis[block_len+2:2*block_len+1])[order])
        push!(peaktable, df)
    end
    return peaktable
end

function visualize_data(data; limit::Int64, clean::Bool, normalise::Bool)
    """
    Make data more legible for heatmap plot
    """
    display_data = data .- minimum(data, dims=1)
    if clean
        display_data, uniform = cleandata(display_data, threshold = 200)
    end
    if normalise
        bin, wavelength_uniformity = cleandata(permutedims(data, (2,1)), threshold=100)
        display_data = display_data ./ maximum(display_data, dims=2)
        display_data[.!wavelength_uniformity,:].=0
    end
    if limit != 0
        display_data[data > limit] = limit
    end

    return display_data
end


function extract_lineactivity(data::Matrix, x_data, wavelength)
    focus = argmin(abs.(x_data .- wavelength))
    linedata = data[focus-10:focus+10, :]
    cleanlinedata, uniformity = cleandata(linedata, threshold=100)
    display(heatmap(cleanlinedata, c=:grays))
    cycles = extractcycles(linedata, uniformity)
    lengths = []
    for c in cycles
        push!(lengths, size(c)[2])
    end
    intensity_over_cycle = zeros(maximum(lengths))
    for c in cycles
        intensity = sum(c, dims=1)
        #intensity_over_cycle = intensity_over_cycle .+ intensity
    end
    display(plot(cycles))
end

function analyse_folder()
    names = readdir("Data", join = true)
    output_folder = "results"
    try readdir(output_folder) catch e mkdir("results") end
    for name in names[5:end]
        data = readdlm(name, ',', Float64, skipstart=25) # some oes data
        x_data = data[:,2]
        data = data[:,5:end]
        cleaned_data, uniformity = cleandata(data)
        display_data = visualize_data(data, limit=0, clean=true, normalise=false)
        display(heatmap(display_data, c=:grays))
        cycles = extractcycleaverages(data, uniformity)
        # extract_lineactivity(data, x_data, 810)
        #peaks_data = peak_analysis_sparse(cycles, x_data, 3, 3)
        #refined_peaks = DataFrame(x0 = vec(peaks_data[1].x0), I_ratio=vec(peaks_data[1].I./peaks_data[2].I), 
        #δx0 = vec(peaks_data[1].x0.-peaks_data[2].x0))
        
        #savefig(output_folder*"/"*filename[1:end-4]*"_heatmap.png")
        #CSV.write(output_folder*"/"*filename[1:end-4]*"peaks.csv", refined_peaks)
        break
    end
end
analyse_folder()